src/utils.rs

```diff
@@ -28,6 +30,60 @@ pub fn is_well_formed_instr(s: &str) -> bool {
     parser::parse::<Instruction>(&buf).is_ok()
 }
 
+/// Decides if a given string is a well-formed text-format Wasm function
+///
+/// Uses wast ParseBuffer to convert string into buffer and wast parser to parse buffer as Module
+/// Encodes Module to binary Wasm and wasmparser parses binary Wasm
+///
+/// # Parameters
+/// s: A string slice representing a Wasm function
+///
+/// # Returns
+/// true: if the function is syntactically well-formed; false otherwise
+pub fn is_well_formed_func(lines: &str) -> bool {
+    //wrap as module
+    let func = "(module\n(func\n".to_string() + lines + "))";
+    //debug: print module
+    println!("{}", func);
+    let buf = match ParseBuffer::new(&func) {
+        Ok(b) => b,
+        Err(_) => {
+            println!("cannot make buffer");
+            return false;
+        }
+    };
+
+    //using wat to get module
+    let wat = match parser::parse::<Wat>(&buf) {
+        Ok(w) => w,
+        Err(_) => return false,
+    };
+    let mut module = match wat {
+        Wat::Module(m) => m,
+        Wat::Component(_) => panic!("No components :("),
+    };
+
+    /*
+    //parse as Module
+    let mut module = match parser::parse::<Module>(&buf) {
+        Ok(m) => m,
+        Err(_) => {
+            println!("cannot parse as module");
+            return false;
+        },
+    };*/
+
+    let bin = match module.encode() {
```

Comment:
These `match` statements could be simplified with syntax like the below -- you might find this more concise (and hopefully comprehensible):

```rust
    let Ok(buf) = ParseBuffer::new(&func) else { return false; };
    let Ok(mut module) = parser::parse::<Module>(&buf) else { return false };
    let Ok(bin) = module.encode() else { return false };
```